### ISSUE TITLE
Performance improvements in Facebook sample

### ISSUE THREAD

**Issue by Nominalista** [eneim/Toro#113]
I was facing with issue which is visible in Facebook feature from sample app. As you certainly noticed sometimes elements in RecyclerView are jumping, because height of items is not fixed. 

Because my goal is to create feed like in Facebook, some of videos are different sizes. So my solution was getting from my video objects info about aspect ratio. Then I set fixed height for video view and because of that my Recycler View items weren't jumping. But the problem is that it lags a little bit and for now I didn't find better solution. But I found something very interesting while exploring real Facebook app. Actually first you see image thumbnail and after that video view is inflating (it looks like, I'm not sure, you can notice that with slower network connection), when you scroll down, it looks like video view is destroying. 

My question is, have you ever consider such solution, or do you have any tips to improve scrolling UX?

**Comment by eneim** [comment]
@Nominalista You are right, loading Video is a huge task, optimized it is expected. I will investigate and update the Facebook Sample to be a more product level sample :D.

**Comment by Nominalista** [comment]
I created sample app which shows my first idea. I set video view height as fixed and everytime I adjust video view height depending on video aspect ratio. [Here](https://github.com/Nominalista/MediaFeed) you can find very simple source code.

I also added the same list but without videos, only thumbnails to see if setting height everytime causes problems (fortunately not, the problem is visible only with videos). 

As I mentioned while fast scrolling view sometimes lags and while slow scrolling view sometimes jumps a little bit. Do you have already in your mind any ideas how to improve performance?

**Comment by eneim** [comment]
@Nominalista thanks a lot for your effort. I think it's caused by the preparation of ExoPlayer everytime the View get attached. At first I try to catch as close as Facebook's performance (they have pretty the same issue actually), but yes it doesn't feel right. I will try to figure out some tune up, or even a better implementation in next update :D.

**Comment by eneim** [comment]
@Nominalista Right now can you take time to try this sample: https://github.com/google/ExoPlayer/pull/2021, it has a same idea but less abstract, but it provide a potential implementation that I may want to adapt :D.

**Comment by Nominalista** [comment]
@eneim thank you for response, I've just tested this example and performance looks a little better, but still it's not FB performance :D Are you able to estimate when you want to work on performance?

**Comment by eneim** [comment]
I may try to put 1 or 2 hours for it each day, right now it gonna be the experiment test. Still have not many ideas :D.

**Comment by Nominalista** [comment]
While looking at demo from ExoPlayer, I noticed one interesting comment, I didn't check if you are doing the same but:

```
// I found that playing on scrolling did not work well without releasing
// and initing a new player for each ViewHolder. My guess is that this is
// to the async nature of the Handler the player uses internally for
// initing and loading. R.Pina 20161018

releasePlayer(); // nulls out player

initializePlayer(); // creates player 
```

**Comment by eneim** [comment]
If you see the PlayerViewHelper class, then there I release and prepare the Player for proper playback.

**Comment by Nominalista** [comment]
I've just changed implementation [here]https://github.com/Nominalista/MediaFeed) using Toro to implementation from sample, which you sent. You can see if it is working faster. Imho it's faster, but Toro onScrollListener is better ;)

**Comment by eneim** [comment]
I tried your repo. Yeah there are some reasons that I would not adapt that implementation (I think I came up with it a few time and gave up on it with same reason :D). I will try to reduce the main thread work load to see the room for improvement.

**Comment by eneim** [comment]
@Nominalista can you checkout [this branch](https://github.com/eneim/Toro/tree/feature/perf-matter-1) and see how you feel. Pretty sure it's been improved, but cannot make sure if it's good practice or not :-?. I also remove the Picasso load to save some main thread resource.

**Comment by Nominalista** [comment]
I am so happy to see news, cause every my solution fails! I will try in an hour, when I get to my computer.

**Comment by Nominalista** [comment]
I noticed much better performance in Facebook sample, but everything cracks when you put in timeline only videos. I mean, I just replace method Factory.newItem() with return new VideoItem() and then you will notice bad performance.

**Comment by Nominalista** [comment]
I don't know if you ever looked at BuzzFeed Video app. They have the best user experience I have ever seen, but something I noticed is that they have only videos in two formats - square and 16:9, so it looks like they have just two seperate view holders. It would only solution for displaying many videos - for every single format new view holder.

**Comment by eneim** [comment]
@Nominalista can you checkout new branch "https://github.com/eneim/Toro/tree/feature/update-video-view" to see if the performance is improved?
